jstz_api/src/text_encoder.rs

```diff
@@ -1,71 +1,249 @@
+    written: u128,
+}
+
+impl TextEncoderEncodeIntoResult {
+    fn try_from_js<'a>(value: &'a JsValue) -> JsResult<GcRefMut<'a, Object, Self>> {
+        value
+            .as_object()
+            .and_then(|obj| obj.downcast_mut::<Self>())
+            .ok_or_else(|| {
+                JsNativeError::typ()
+                    .with_message("Failed to convert js value into rust type `TextEncoderEncodeIntoResult`")
+                    .into()
+            })
+    }
+}
+
+impl Default for TextEncoderEncodeIntoResult {
+    fn default() -> TextEncoderEncodeIntoResult {
+        TextEncoderEncodeIntoResult {
+            read: 0,
+            written: 0,
         }
-        let str = data.to_std_string_escaped();
-        let encoded = DEFAULT_ENGINE.decode(&str).map_err(on_err)?;
-        let encoded_str = core::str::from_utf8(encoded.as_slice()).map_err(on_err)?;
+    }
+}
 
-        Ok(encoded_str.into())
+impl NativeClass for TextEncoderEncodeIntoResult {
+    type Instance = TextEncoderEncodeIntoResult;
+
+    const NAME: &'static str = "TextEncoderEncodeIntoResult";
+
+    fn constructor(
+        _this: &JsNativeObject<TextEncoderEncodeIntoResult>,
+        _args: &[JsValue],
+        _: &mut Context<'_>,
+    ) -> JsResult<TextEncoderEncodeIntoResult> {
+        Ok(TextEncoderEncodeIntoResult::default())
     }
-    fn btoa(data: &JsString) -> JsResult<JsString> {
-        let str = data.to_std_string_escaped();
-        let encoded = DEFAULT_ENGINE.encode(&str);
-        Ok(encoded.into())
+
+    fn init(class: &mut ClassBuilder<'_, '_>) -> JsResult<()> {
+        let read = accessor!(
+            class.context(),
+            TextEncoderEncodeIntoResult,
+            "read",
+            get:((x, _context) => Ok(JsBigInt::new(x.read).into()))
+        );
+        let written = accessor!(
+            class.context(),
+            TextEncoderEncodeIntoResult,
+            "written",
+            get:((x, _context) => Ok(JsBigInt::new(x.written).into()))
+        );
+        class
+            .accessor(js_string!("read"), read, Attribute::all())
+            .accessor(js_string!("written"), written, Attribute::all());
+
+        Ok(())
     }
 }
 
+// Note: Using Box as a way to get a fixed size array, probably wrong
+type Uint8Array = Box<[u8]>;
+
+impl TextEncoder {
+    fn try_from_js<'a>(value: &'a JsValue) -> JsResult<GcRefMut<'a, Object, Self>> {
+        value
+            .as_object()
+            .and_then(|obj| obj.downcast_mut::<Self>())
+            .ok_or_else(|| {
+                JsNativeError::typ()
+                    .with_message(
+                        "Failed to convert js value into rust type `TextEncoder`",
+                    )
+                    .into()
+            })
+    }
+
+    fn encoding() -> &'static str {
+        "utf-8"
+    }
+
+    fn encode(input: Option<String>) -> Uint8Array {
+        let s = input.unwrap_or_else(|| "".to_string());
+        s.into_bytes().into_boxed_slice()
+    }
```

Comment:
Definitely not spec compliant 😅 

I think in terms of sharing code between `encode` and `encode_into`, it would be best to implement their process: https://encoding.spec.whatwg.org/#concept-encoding-process.

Reply:
I struggle a lot to understand the spec..

> A [handler](https://encoding.spec.whatwg.org/#handler) algorithm takes an input [I/O queue](https://encoding.spec.whatwg.org/#concept-stream) and an [item](https://infra.spec.whatwg.org/#list-item), and returns finished, one or more [items](https://infra.spec.whatwg.org/#list-item), error optionally with a [code point](https://infra.spec.whatwg.org/#code-point), or continue.

what's the handler here ?